flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DateTime.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.dataformat;
+
+import org.apache.flink.util.Preconditions;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+
+/**
+ * Immutable SQL TIMESTAMP type with nanosecond precision.
+ *
+ * <p>This class is composite of a millisecond and nanoOfMillisecond. The millisecond part
+ * holds the integral second and the milli-of-second. The nanoOfMillisecond holds the
+ * nano-of-millisecond, which should between 0 - 999_999.
```

Comment:
Since `java.sql.Timestamp` stores millisecond in its nanos part, does it mean this class has different ranges from `java.sql.Timestamp`?

Reply:
No. They have the same range. The differences is:
`java.sql.Timestamp` holds integral second plus nanosecond(with milli-part), and
`DateTime` holds integral second(with milli-part) plus nanoOfMillisecond.

Reply:
I think the max value of this class is `Long.MAX_VALUE (ms) + 999_999 (ns)`, while the max value of `java.sql.Timestamp` is `Long.MAX_VALUE/1000*1000 (ms) + 999_999_999 (ns)`. So java's Timestamp has a bigger max value, no?

Reply:
There are 3 methods to obtain a  `java.sql.Timestamp`
1) Timestamp.valueOf("yyyy-mm-dd hh:mm:ss[.fffffffff]"), the  maximum one is 9999-12-31 23:59:59.999999999". It's much smaller than Long.MAX

2) new Timestamp(long), the maximum one is new Timestamp(Long.MAX), and the long part hold 9223372036854775000, the int part hold 807000000. We will put the `807` to long part, which is exactly the `Long.MAX`. 

3) Timestamp.valueOf(LocalDateTime): LocalDateTime has bigger range than `java.sql.Timestamp` (also bigger then `java.time.Instant`), ignore this.

So we can represent the biggest `java.sql.Timestamp`, which is "292278994-08-17 07:12:55.807" in human view.

Reply:
Actually I can do things like `Timestamp.valueOf(LocalDateTime.of(292278994, 8, 17, 15, 12, 55, 999999999))` and the resulting timestamp instance is able to hold this bigger value.
But I suppose this difference is acceptable since the extreme case doesn't seem very useful in normal scenarios.

Reply:
`Timestamp.valueOf(LocalDateTime.of(292278994, 8, 17, 15, 12, 55, 999999999))`.toString
returns "292269055-12-03 00:47:03.999999999" on UTC+8. It's overflow？